include/uv/darwin.h

```diff
@@ -30,9 +30,25 @@
 # define UV_PLATFORM_SEM_T semaphore_t
 #endif
 
+#if defined(_USE_LIBINFO)
+# define UV_IO_PRIVATE_PLATFORM_EXTRA_FIELDS                                  \
+  void *data;
+
+# define UV_GETADDRINFO_PLATFORM_PRIVATE_EXTRA_FIELDS                         \
+    uv__io_t async_addrinfo_io;
+# define UV_GETNAMEINFO_PLATFORM_PRIVATE_EXTRA_FIELDS                         \
+    uv__io_t async_getnameinfo_io;
```

Comment:
@bnoordhuis here you commented on the extension of `uv_getnameinfo_t/uv_getaddrinfo_t`, but I guess you meant the one on top `UV_IO_PRIVATE_PLATFORM_EXTRA_FIELDS` which is a `void *data`, am I correct?

Reply:
It's applies to all three, they all change the ABI (the size of the structs they're used in.)

Reply:
ouch, then it becomes a major issue... when is it ok to force the users to recompile (that would be enough)?

another approach (but then it makes the handle of "fallback" even harder) is to use the `struct uv__work work_req` space (using an `union` would make it more future proof):

```c
struct uv__io_s {
  uv__io_cb cb; // ptr
  void* pending_queue[2]; // 2 ptr
  void* watcher_queue[2]; // 2 ptr
  unsigned int pevents; // 4 bytes
  unsigned int events;  // 4 bytes
  int fd; // 4 bytes
}; //  5 ptr + 12 bytes = 32 or 52 bytes

struct uv__work {
  void (*work)(struct uv__work *w); // ptr
  void (*done)(struct uv__work *w, int status); // ptr
  struct uv_loop_s* loop; // ptr
  void* wq[2]; // 2 ptr
}; // total: 5 ptr = 20 or 40 bytes

// since sizeof(uv__io_s) > sizeof(uv_work), we must do dynamic allocation:

struct uv_getaddrinfo_libinfo_io {
  void *_zero; // uv__work.work == NULL could be used to check what's in use
  uv__io_t *io; // must be allocated as it won't fit into struct uv__work
  void *io_data; // will be the container since we cannot use container_of() anymore :-(
}; // total: 3 ptr = 12 or 24 bytes


#define UV_GETADDRINFO_PRIVATE_FIELDS                                         \
  union { \
     struct uv__work work_req; \
     struct uv_getaddrinfo_libinfo_io io_req; \
  } \
// just to illustrate, would need macros to be properly defined, only on macOS with the libinfo enabled
```

then if `work_req.work == NULL` we're using it as an async load and we must handle it as `io_req`. Since `uv__io_t` is too big, we need to dynamically allocate it... but there is no need to keep an external data structure to point back to it (hashes or the likes).

Ugly, but should work if we can't ask people to recompile.

Reply:
I guess he meant to remove all additions and only use `loop->internal_fields`. However I have a question for you @bnoordhuis I was looking at the this field and it is already populated right here https://github.com/libuv/libuv/blob/4296fec7f50145e2a307f3db7ae22984713976a7/src/unix/loop.c#L40-L43

Did you mean to add a hash table in here  (at struct uv__loop_internal_fields_s) ?
https://github.com/libuv/libuv/blob/5c85d67bc781430f2db8f446795494aa8d399de8/src/uv-common.h#L368-L370

Reply:
> Did you mean to add a hash table in here (at struct uv__loop_internal_fields_s) ?

That's right.

> when is it ok to force the users to recompile

When libuv 2.0 is released. :-)

Reply:
@bnoordhuis is it okay to go with my suggested approach? It removes a lot of work to manage a hashtable and doesn't break the ABI. We can add the union or simply keep it internal with `#if sizeof(x) < sizeof(y)` checks to compile-time check there is no misusage.

TL;DR is `uv__work` can fit the pointers to a new memory block (which would be needed in a hash table as well, but no hashing/lookup). To differentiate from a regular `uv__work` and this new memory block container we'd use the first pointer (`uv__work->work`), if it's null it's used by the async/libinfo work and the other member will point to the newly allocated memory. As it's not expected to have `uv__work->work == NULL`, then it should be enough and not break anything else. Alignments and all are also preserved.